utils/vote_item_builder.py

```diff
@@ -0,0 +1,198 @@
+
+class VoteItem:
+    """Factory for creating different types of vote actions."""
+
+    @staticmethod
+    def agent(target_address: str, encoded_data: str) -> VoteItemDict:
+        """Create vote action that will be executed through the Agent."""
+        return {"type": ActionType.AGENT, "address": target_address, "data": encoded_data}
+
+    @staticmethod
+    def admin(target_address: str, encoded_data: str) -> VoteItemDict:
+        """Create vote action that will be executed through Dual Governance."""
+        return {"type": ActionType.DUAL_GOVERNANCE_ADMIN, "address": target_address, "data": encoded_data}
+
+    @staticmethod
+    def voting(target_address: str, encoded_data: str) -> VoteItemDict:
+        """Create vote action that will be executed directly by the Voting contract."""
+        return {"type": ActionType.DIRECT_VOTING, "address": target_address, "data": encoded_data}
+
+# ──────────────────────────────────────────────────────────────────────────
+#  Call Script Encoders
+# ──────────────────────────────────────────────────────────────────────────
+
+def encode_agent_forward_call(call_script: Sequence[Tuple[str, str]]) -> ExecutableCall:
+    """Encode a call script for Agent forwarding."""
+    agent = contracts.agent
+    print(f"Encoding agent forward call for script: {call_script}")
+    return (AGENT, agent.forward.encode_input(encode_call_script(call_script)))
+
+# ──────────────────────────────────────────────────────────────────────────
+#  Vote Items Validation and Aggregation
+# ──────────────────────────────────────────────────────────────────────────
+
+def validate_and_aggregate_vote_items(
+    vote_items: List[VoteItem]
+) -> Tuple[List[str], List[VoteItemDict]]:
+    """
+    Validates vote items ordering and separates descriptions from actions.
+
+    Validation rule: Direct voting actions must appear either before all other
+    actions or after all other actions (not interleaved).
+    """
+    descriptions: List[str] = []
+    actions: List[VoteItemDict] = []
+    action_types: List[str] = []
+
+    for description, action in vote_items:
+        descriptions.append(description)
+        actions.append(action)
+        action_types.append(action["type"])
+
+    # Find the boundaries of non-voting actions
+    first_non_voting_index = next(
+        (i for i, action_type in enumerate(action_types) if action_type != ActionType.DIRECT_VOTING),
+        len(action_types)
+    )
+    last_non_voting_index = (
+        len(action_types) - 1 - next(
+            (i for i, action_type in enumerate(reversed(action_types)) if action_type != ActionType.DIRECT_VOTING),
+            len(action_types)
+        )
+    )
+
+    # Check if voting actions are interleaved with non-voting actions
+    interleaved_voting_actions = action_types[first_non_voting_index:last_non_voting_index + 1]
+    if any(action_type == ActionType.DIRECT_VOTING for action_type in interleaved_voting_actions):
```

Comment:
Scenario example:
1) Agent grants the permission to voting
2) Voting does some actions
3) Agent revokes the permission from voting